### PR TITLE
feat: fix IPv6 url serialization, use node 18

### PR DESCRIPTION
**Pull request by milesrichardson**
This PR upgrades the development environment to Node v18 (which will [replace v16 as Active LTS in October](https://nodejs.org/en/about/releases/)) and adds CI tests for Node v14, v16 and v18. It fixes the bugs in Node v18 affecting downstream package `mswjs/interceptors` described in https://github.com/mswjs/interceptors/issues/282 and patched in https://github.com/mswjs/interceptors/pull/283.

I believe this also closes #8 (fixes the underlying bug so there is no need for it).

This PR does not update the version for release, since I'm not sure the process for that. Once you merge it, and release a new version of the `page-with` package, you can update https://github.com/mswjs/interceptors/pull/283 to revert [the commit that patches over this bug](https://github.com/mswjs/interceptors/pull/283/commits/4291844df928b0f78323574df83d0e9d2e0b7cb6). You could also revert [the commit pinning memfs](https://github.com/mswjs/interceptors/pull/283/commits/1c4309e190078ec8ef0a686b7b72ef7c804ff407), since that will be a no-op once this package depends on the latest version of it.

See commits for more details. The bugfix is in 3e48ce2e21c58324676a6df5dc67281cb243dc97 and the rest of the commits are fixing compatibility issues and resolving some warnings. 

AFAICT, this PR contains no breaking changes.

### CODE REVIEW

**Comment by milesrichardson** [comment]
Note: see my fork for example of successful test run: https://github.com/milesforks/page-with/pull/1

File: src/server/PreviewServer.ts
Comment (kettanaito) [inline src/server/PreviewServer.ts]: Nitpick: a rather complex expression. Consider:

```js
const host = family === 'IPv6' ? `[${address}]` : address
const url = `http://${host}:${port}`
```

File: src/server/PreviewServer.ts
Comment (kettanaito) [inline src/server/PreviewServer.ts:122]: Who/how is this `family` be consumed further down the chain?

**Review by kettanaito — APPROVED** [review]
Thanks for seeing this through, @milesrichardson! 

I think the changes are good. I've left one nitpick but leave it up to you to resolve it or not (just ping me in any case so we could merge this).

File: src/server/PreviewServer.ts
Comment (milesrichardson) [reply src/server/PreviewServer.ts]: Good idea! I've implemented your suggestion, except I named the variable "`serializedAddress`," which seems more contextually appropriate, especially given the existing `host` variable in the outer scope, set by the second argument to `listen` (default `"localhost"`).

I modified the existing commit and force pushed the change. Assuming the tests pass, I think this PR is ready to merge.

File: src/server/PreviewServer.ts
Comment (milesrichardson) [reply src/server/PreviewServer.ts:122]: Ironically, the code that could benefit from it in https://github.com/mswjs/interceptors/pull/283 is the patch to `page-info` that won't be necessary once this is merged 😸 I think the patch for `@open-draft/test-server` could also use it, but I don't remember the exact code path. 

Basically, any logic that needs to differentiate between IPv6 and IPv4 based on the URL currently "guesses" with logic like `isIPv6 = address.includes(":")`. So returning the family here simply preserves information that we already have available from the `AddressInfo` object returned by `listen`. 

It's not strictly necessary but makes life easier downstream and shouldn't break anything. I guess it could break somebody's snapshot tests or something, but that would be surprising (especially since `this.connectionInfo` is never actually _returned_ - in the patch, I accessed it by exporting the mutable `server` in the browser integration tests setup function).

If you want to remove this, feel free to revert this commit (95f78f00df7b0546214a6dfe12647d7b3254e452) (or let me know your decision and I can force push the change).

**Comment by milesrichardson** [comment]
Thanks for the review! I addressed your nitpick and answered your question. :) 

Since I force pushed the change, you need to approve the workflow run again, but assuming all tests pass, then it's ready to merge IMO

**Review by kettanaito — APPROVED** [review]

**Comment by kettanaito** [comment]
Thank you for making this library better, @milesrichardson! 🎉 

I will introduce this change in `test-server` as well.

**Comment by milesrichardson** [comment]
Awesome, thanks! Let me know when the new npm package is released and I'll swap out the patch from https://github.com/mswjs/interceptors/pull/283

p.s. I'm very close to having `msw` itself fully migrated to Node 18 and intercepting native fetch by default. I'll open a PR when it's ready but you can follow here https://github.com/milesforks/msw/pull/2

tl;dr Adding the `FetchInterceptor` is easy. Upgrading to Node v18 so that we can actually test its behavior is not so easy - basically due to IPv6 issues (which I suppose would be present in earlier versions too). I've fixed those though and am now working on the last few errors

**Comment by kettanaito** [comment]
@milesrichardson, I've just released 0.6.0. Thank you for making this happen.

File: src/server/PreviewServer.ts
Comment (kettanaito) [reply src/server/PreviewServer.ts:122]: @milesrichardson, thanks for all your work on this. I'm also addressing related things in https://github.com/mswjs/interceptors/pull/354.

Comment (milesrichardson) [reply src/server/PreviewServer.ts:122]: Awesome! Sorry for all the back and forth (or no forth, as the case may be :eyes:). Let me know if I can help.

I haven't tried the latest version of msw in my Node 18 project yet (still using the hacky tarball I built), but I'll try switching it soon. (btw, that project is now open source, if you want to see how msw is used there: https://github.com/splitgraph/madatdata)

Comment (milesrichardson) [reply src/server/PreviewServer.ts:122]: Last I remember, everything was good to go. There were some failing tests, but the reason was because of a memory leak in the test runner, and the solution (which I didn't implement) would be to split it up into two calls to the test runner (I couldn't fix the memory leak after two days of trying).

I'm assuming you won't be using the code I wrote directly, so maybe that won't matter to you, but you can at least be assured that the latest commit in each respective project was known to be working at least once. Feel free to copy paste 😅

Comment (milesrichardson) [reply src/server/PreviewServer.ts:122]: I think this diff should contain the few necessary fixes to IPv6 serialization in msw-interceptors: https://github.com/milesforks/msw-interceptors/pull/6/files

And this is where the bulk of the work on `msw` was. This is the one where tests are only failing because of the memory leak, but otherwise everything else works. There is some IPv6 stuff in here too: https://github.com/milesforks/msw/pull/2/files

The commit messages are pretty detailed, e.g. https://github.com/milesforks/msw/pull/2/commits/b57db77af99ca1d3cc10a48bf44f3d772c2eacca ("fix(coercePath): widen regex to escape : in IPv6 address as well")

**Pull request merged**
